### PR TITLE
Add a DNS client example.

### PR DESCRIPTION
**Pull request by seedeed**
Motivation:

It seems that there is no DNS client example in Netty project so far.

Modification:

Add a Netty DNS client example.

### CODE REVIEW

**Comment by seedeed** [comment]
@hyperxpro done.

**Comment by hyperxpro** [comment]
Seems good.

cc @normanmaurer

File: example/src/main/java/io/netty/example/dns/udp/DnsClient.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            if (record.type() == DnsRecordType.A) {
+                //just print the IP after query
+                DnsRawRecord raw = (DnsRawRecord) record;
+                System.out.println(NetUtil.bytesToIpAddress(ByteBufUtil.getBytes(raw.content())));
+            }
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
+        InetSocketAddress addr = SocketUtils.socketAddress(DNS_SERVER_HOST, DNS_SERVER_PORT);
```
Comment (normanmaurer) [inline example/src/main/java/io/netty/example/dns/udp/DnsClient.java:69]: don't use `SocketUtils` in an example as its a internal class.

File: example/src/main/java/io/netty/example/dns/udp/DnsClient.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                        }
+                    });
+                 }
+             });
+            Channel ch = b.bind(0).sync().channel();
+            DnsQuery query = new DatagramDnsQuery(null, addr, 1).setRecord(
+                    DnsSection.QUESTION,
+                    new DefaultDnsQuestion(QUERY_DOMAIN, DnsRecordType.A));
+            ch.writeAndFlush(query);
+            latch.await(10L, TimeUnit.SECONDS);
```
Comment (normanmaurer) [inline example/src/main/java/io/netty/example/dns/udp/DnsClient.java:100]: you could also just call `ch.closeFuture().sync()` and call `ch.close()` in your `channelRead0` method

File: example/src/main/java/io/netty/example/dns/udp/DnsClient.java [hunk 3]
```diff
@@ -81,24 +81,32 @@ protected void initChannel(DatagramChannel ch) throws Exception {
...
                      .addLast(new SimpleChannelInboundHandler<DatagramDnsResponse>() {
                         @Override
-                        protected void channelRead0(ChannelHandlerContext ctx, DatagramDnsResponse msg)
-                                throws Exception {
+                        protected void channelRead0(ChannelHandlerContext ctx, DatagramDnsResponse msg) {
                             try {
                                 handleQueryResp(msg);
                             } finally {
-                                latch.countDown();
+                                ctx.channel().close();
```
Comment (normanmaurer) [inline example/src/main/java/io/netty/example/dns/udp/DnsClient.java:88]: nit: ctx.close();

**Comment by seedeed** [comment]
@normanmaurer PTAL

File: example/src/main/java/io/netty/example/dns/udp/DnsClient.java [hunk 4]
```diff
@@ -81,24 +81,32 @@ protected void initChannel(DatagramChannel ch) throws Exception {
...
+            ScheduledFuture<?> f = ch.eventLoop().schedule(new Runnable() {
+                @Override
+                public void run() {
+                    if (ch.isActive()) {
+                        System.err.println("dns query timeout.");
+                        ch.close();
+                    }
+                }
+            }, 10L, TimeUnit.SECONDS);
+            ch.closeFuture().sync();
```
Comment (normanmaurer) [inline example/src/main/java/io/netty/example/dns/udp/DnsClient.java:108]: just remove the `schedule` and call `ch.closeFuture().await(....)`

**Comment by seedeed** [comment]
@normanmaurer done. PTAL

File: example/src/main/java/io/netty/example/dns/udp/DnsClient.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+             });
+            final Channel ch = b.bind(0).sync().channel();
+            DnsQuery query = new DatagramDnsQuery(null, addr, 1).setRecord(
+                    DnsSection.QUESTION,
+                    new DefaultDnsQuestion(QUERY_DOMAIN, DnsRecordType.A));
+            ch.writeAndFlush(query).sync();
+            boolean succ = ch.closeFuture().await(10, TimeUnit.SECONDS);
+            if (!succ) {
+                System.err.println("dns query timeout!");
+                if (ch.isActive()) {
```
Comment (normanmaurer) [inline example/src/main/java/io/netty/example/dns/udp/DnsClient.java:101]: you can remove this `if`. It doesn't harm to close if it was not active

**Comment by seedeed** [comment]
done

**Comment by normanmaurer** [comment]
@seedeed thanks a lot

**Pull request merged**